### PR TITLE
fix: Adapt the Secrets plugin API to use kubernetes secrets

### PR DESCRIPTION
**Pull request by vinokurig**
Signed-off-by: Igor Vinokur <ivinokur@redhat.com>

### What does this PR do?
Rework the Secrets API from upstream theia to use kubernetes secrets

### What issues does this PR fix or reference?

fixes https://github.com/eclipse/che/issues/19837

### How to test this PR?

1. Start a workspace from devfile:
```
apiVersion: 1.0.0
metadata:
  name: wksp-xhxvu
components:
  - type: cheEditor
    reference: 'https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-1166/che_theia_meta.yaml'
  - type: chePlugin
    reference: 'https://pastebin.com/raw/riqiReMi'
```
2. Run Test Get Password command and see a notification with empty (undefined) password.
3. Run Test Set Password command and see a notification about password change.
4. Run Test Get Password command and see a notification with the updated password.
5. Run Test Delete Password command and see a notification about password change.
6. Run Test Get Password command and see a notification with empty (undefined) password.

P.S. This will work only for newly created namespace, see https://github.com/eclipse-che/che-server/pull/44#pullrequestreview-699942442
### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable

### CODE REVIEW

**Comment by vinokurig** [comment]
Depends on https://github.com/eclipse-che/che-server/pull/44

**Comment by dmytro-ndp** [comment]
[crw-ci-test --rebuild]

**Review by vzhukovs — APPROVED** [review]
lgtm 👍

File: extensions/eclipse-che-theia-credentials/src/node/che-credentials-server.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import * as k8s from '@kubernetes/client-node';
+
+import { CredentialsServer, PasswordContent } from '../common/credentials-protocol';
+import { inject, injectable } from 'inversify';
+
+import { CheK8SServiceImpl } from '@eclipse-che/theia-remote-impl-che-server/lib/node/che-server-k8s-service-impl';
+import { CheServerWorkspaceServiceImpl } from '@eclipse-che/theia-remote-impl-che-server/lib/node/che-server-workspace-service-impl';
+
+@injectable()
+export class CheCredentialsServer implements CredentialsServer {
```
Comment (benoitf) [inline extensions/eclipse-che-theia-credentials/src/node/che-credentials-server.ts:20]: For this classe, we should rely on the mounted files for reading secrets.

About delete and create it should be possible to do it using K8S API

**Review by azatsarynnyy — COMMENTED** [review]
If I understand correctly, with this PR, we don't need `keytar`-based code anymore, e,g, [KeytarServiceImpl](https://github.com/eclipse-theia/theia/blob/9578b84f8a2d8fa7a89c0544690c8e656cb08e30/packages/core/src/node/keytar-server.ts#L29) or [KeytarCredentialsProvider](https://github.com/eclipse-theia/theia/blob/9578b84f8a2d8fa7a89c0544690c8e656cb08e30/packages/core/src/browser/credentials-service.ts#L83).
Can we unbind the unused keytar-related components in Che-Theia to avoid including `libsecret` into the image?

As recently, we've fixed starting Che-Theia backend by including `libsecret` in https://github.com/eclipse-che/che-theia/commit/07aeab7c143dc88d0a972ab32c6dc82b37cbc9a6
image

File: extensions/eclipse-che-theia-credentials/src/node/che-credentials-server.ts (hunk 1 above)
Comment (vinokurig) [reply extensions/eclipse-che-theia-credentials/src/node/che-credentials-server.ts:20]: > About delete and create it should be possible to do it using K8S API

Could you please elaborate more on this?

File: extensions/eclipse-che-theia-credentials/src/browser/che-credentials-service.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
+/**********************************************************************
+ * Copyright (c) 2019-2020 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-credentials/src/browser/che-credentials-service.ts]: done

File: extensions/eclipse-che-theia-credentials/src/browser/credentials-frontend-module.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+/**********************************************************************
+ * Copyright (c) 2020 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-credentials/src/browser/credentials-frontend-module.ts]: done

File: extensions/eclipse-che-theia-credentials/src/node/che-theia-credentials-backend-module.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+/**********************************************************************
+ * Copyright (c) 2019-2020 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-credentials/src/node/che-theia-credentials-backend-module.ts]: done

**Comment by vinokurig** [comment]
@azatsarynnyy 
> If I understand correctly, with this PR, we don't need keytar-based code anymore, e,g, KeytarServiceImpl or KeytarCredentialsProvider.
Can we unbind the unused keytar-related components in Che-Theia to avoid including libsecret into the image?
>
>As recently, we've fixed starting Che-Theia backend by including libsecret in 07aeab7

When I unbind the KeytarServiceImpl and KeytarCredentialsProvider classes the che-theia build fails with an error:
```
#21 403.7     'ERROR in ../../node_modules/keytar/build/Release/keytar.node 1:0\n' +
#21 403.7     "Module parse failed: Unexpected character '' (1:0)\n" +
#21 403.7     'You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders\n' +
#21 403.7     '(Source code omitted for this binary file)\n' +
#21 403.7     ' @ ../../node_modules/keytar/lib/keytar.js 1:13-52\n' +
#21 403.7     ' @ ../../packages/core/lib/node/keytar-server.js 63:13-30\n' +
#21 403.7     ' @ ../../che-theia/extensions/eclipse-che-theia-credentials/lib/browser/credentials-frontend-module.js 16:22-67\n' +
#21 403.7     ' @ ./src-gen/frontend/index.js 100:47-128\n' +
#21 403.7     '\n' +
#21 403.7     'webpack 5.46.0 compiled with 1 error and 75 warnings in 140062 ms\n' +
#21 403.7     '\n' +
#21 403.7     'info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.\n',
```
Probably we can fix it as a next iteration.

**Comment by vinokurig** [comment]
@azatsarynnyy @benoitf Since the [server-side part](https://github.com/eclipse-che/che-server/pull/68) is merged I am going to merge this tomorrow if no more objections.

**Comment by azatsarynnyy** [comment]
> @azatsarynnyy @benoitf Since the [server-side part](https://github.com/eclipse-che/che-server/pull/68) is merged I am going to merge this tomorrow if no more objections.

Thanks @vinokurig ! The PR looks good to me :+1: 

But I'm worried about removing the `keytar` dependency, since we don't use it in Che-Theia, and we have to provide it with all our Che/CRW images.

> ...
> Probably we can fix it as a next iteration.

It's a good idea. It would be great to create an issue for that and at least investigate it during one of the next sprints.

**Pull request merged**